internal/codeintel/autoindexing/internal/store/store_indexes.go

```diff
@@ -472,16 +550,67 @@ func (s *store) IsQueued(ctx context.Context, repositoryID int, commit string) (
 	}})
 	defer endObservation(1, observation.Args{})
 
-	count, _, err := basestore.ScanFirstInt(s.db.Query(ctx, sqlf.Sprintf(isQueuedQuery, repositoryID, commit, repositoryID, commit)))
-	return count > 0, err
+	isQueued, _, err := basestore.ScanFirstBool(s.db.Query(ctx, sqlf.Sprintf(
+		isQueuedQuery,
+		repositoryID, commit,
+		repositoryID, commit,
+	)))
+	return isQueued, err
 }
 
 const isQueuedQuery = `
-SELECT COUNT(*) WHERE EXISTS (
-	SELECT id FROM lsif_uploads_with_repository_name WHERE repository_id = %s AND commit = %s AND state NOT IN ('deleted', 'deleting')
-	UNION
-	SELECT id FROM lsif_indexes_with_repository_name WHERE repository_id = %s AND commit = %s
-)
+SELECT
+	EXISTS (
+		SELECT 1
+		FROM lsif_uploads u
+		WHERE
+			repository_id = %s AND
+			commit = %s AND
+			state NOT IN ('deleting', 'deleted') AND
+			associated_index_id IS NULL
+	)
+
+	OR
+
+	EXISTS (
+		SELECT 1
+		FROM (
+			SELECT DISTINCT ON (root, indexer) should_reindex
+			FROM lsif_indexes
+			WHERE repository_id = %s AND commit = %s
+			ORDER BY root, indexer, queued_at DESC
+		) _
+		WHERE NOT should_reindex
+	)
+`
+
+// IsQueuedRootIndexer returns true if there is an index or an upload for the given (repository, commit, root, indexer).
+func (s *store) IsQueuedRootIndexer(ctx context.Context, repositoryID int, commit string, root string, indexer string) (_ bool, err error) {
+	ctx, _, endObservation := s.operations.isQueued.With(ctx, &err, observation.Args{LogFields: []log.Field{
+		log.Int("repositoryID", repositoryID),
+		log.String("commit", commit),
+		log.String("root", root),
+		log.String("indexer", indexer),
+	}})
+	defer endObservation(1, observation.Args{})
+
+	isQueued, _, err := basestore.ScanFirstBool(s.db.Query(ctx, sqlf.Sprintf(isQueuedRootIndexerQuery, repositoryID, commit, root, indexer)))
+	return isQueued, err
+}
+
+const isQueuedRootIndexerQuery = `
+SELECT
+	EXISTS (
+		SELECT 1
+		FROM lsif_indexes
+		WHERE
+			repository_id  = %s AND
+			commit         = %s AND
+			root           = %s AND
+			indexer        = %s AND
+			should_reindex = false
+		ORDER BY queued_at DESC
+	)
```

Comment:
```suggestion
SELECT
	SELECT true
	FROM lsif_indexes
	WHERE
		repository_id  = %s AND
		commit         = %s AND
		root           = %s AND
		indexer        = %s AND
		should_reindex = false
```

I think this should work with the same `ScanFirstBool` call above (it will be false if there are no rows returned). The order by in the the original query didn't seem to do anything.

Reply:
I think I forgot a `LIMIT 1`. We want it to return whether or not the most recently-queued index is marked for reindexing, right?

Reply:
```suggestion
SELECT NOT should_reindex
FROM lsif_indexes
WHERE
	repository_id  = %s AND
	commit         = %s AND
	root           = %s AND
	indexer        = %s AND
ORDER BY queued_at DESC
LIMIT 1
```

Oh yes, you're right. I think you want to return the `should_reindex` field explicitly for that one record in that case.